### ISSUE TITLE
SIGSEGV issue for BFTKmer

### ISSUE THREAD

**Issue by kalebo** [kalebo/go-bft#1]
Automatic garbage collection of the c backed structs is currently achieved by wrapping them in golang structs and specifying a finalizer for them. This seems to work most of the time, however after traversing a small simulated sequence an inconsistent SIGSEGV appears. It is associated with the freeing of the BFT_kmer cstruct. 

Freeing the contents of the cstruct works as expected. However, removing the contents, and freeing the cstruct itself causes the SIGSEGV. The cstruct was [malloc'd by libbft](https://github.com/GuillaumeHolley/BloomFilterTrie/blob/2f5cbd6002b6fff74ac1c104f0955ccd4736f073/src/bft.c#L961) and then ownership of the struct should be moved to the caller of get_successors() on return.

The code in question is seen below: https://github.com/kalebo/go-bft/blob/cace6fb1ad9f88e41aa3f6ca83345b0f3f4b57a3/bftkmer.go#L38-L46

Below is an example program that reproduces the error:
```go
package main

import (
	"fmt"
	"runtime"
	"time"

	"github.com/kalebo/go-bft"
)

func example() {
	graph := bft.NewBFTGraph("out.bft")
	fmt.Println("Loaded BFT...")

	kmer := graph.GetKmer("CCGCAAGCT")
	fmt.Print("Initial: ")
	fmt.Println(kmer)

	fmt.Println("Successors...")
	for _, n := range kmer.GetSuccessors() {
		fmt.Println(n)
	}

	fmt.Println("Following first...")
	followFirst(kmer, 0)

	graph.Free()
}

func followFirst(kmer *bft.BFTKmer, i int) {
	succ := kmer.GetSuccessors()
	if len(succ) > 0 && i < 100 {
		fmt.Println("next:", succ[0])
		followFirst(succ[0], i+1)
	}
}

func main() {
	example()
	runtime.GC()
	time.Sleep(3 * time.Second)
	println("Allocated:", bft.Alloc, "; Freed:", bft.Freed)
}
```

When running the above with the contents of 
[out.bft.zip](https://github.com/kalebo/go-bft/files/1899630/out.bft.zip) the typical number of successful frees is 145 (the total number of BFTKmer allocations) or a number very close to ether 123 or 86 when crashing at a segfault. Over a total of 25 crashes only 2 BFTKmers had kmers that were not unique. The fact that most crashes are at 123, 122, or 86, 85 seems significant, but I'm not sure how.

## TL;DR:
SIGSEGVs are not random but they are inconsistent.

**Comment by kalebo** [comment]
Duh! I know what's going wrong! I'm trying to free only part of an array and that's not how you C. Effectively, I can only free the pointer that is returned from malloc (and friends) which makes total sense when you think about how malloc works!  Therefore the fix to this issue is to either memcpy the cstructs out of the array of 4 and then free the array, or to coordinate the finalization of all BFTKmers that point into the same array.

**Comment by kalebo** [comment]
Issue fixed. Note that the GC needs three iterations to fully collect all objects (kmers, array containers and graph) due to how finalizers are implemented. See the godocs for [SetFinalizer](https://golang.org/pkg/runtime/#SetFinalizer) for more.

The above can be verified with the following example program:

```go

package main

import (
        "fmt"
        "runtime"
        "time"

        "github.com/kalebo/go-bft"
)

func example() {
        graph := bft.NewBFTGraph("out.bft")
        fmt.Println("Loaded BFT...")

        kmer := graph.GetKmer("CCGCAAGCT")
        fmt.Print("Initial: ")
        fmt.Println(kmer)

        fmt.Println("Successors...")
        for _, n := range kmer.GetSuccessors() {
                fmt.Println(n)
        }

        fmt.Println("Following first...")
        followFirst(kmer, 0)
}

func followFirst(kmer *bft.BFTKmer, i int) {
        succ := kmer.GetSuccessors()
        if len(succ) > 0 && i < 100 {
                fmt.Println("next:", succ[0])
                followFirst(succ[0], i+1)
        }
}

func main() {
        example()
        runtime.GC() // collects the kmers
        time.Sleep(time.Second)
        println("Allocated:", bft.Alloc, "; Freed:", bft.Freed)

        runtime.GC() // will collect the containing arrays
        time.Sleep(time.Second)
        println("Allocated:", bft.Alloc, "; Freed:", bft.Freed)

        runtime.GC() // will collect the graph
        time.Sleep(time.Second)
        println("Allocated:", bft.Alloc, "; Freed:", bft.Freed)
}

```
